### DISCUSSION TITLE
解释最近代码的改动

### DISCUSSION THREAD

**Discussion by JohanChane** [JohanChane/clashtui#32] [General]
最近添加的功能:
-   配置增加 timeout 字段。因为下载某些资源时, 可能网络比较差, 所以让用户自己设置超时时间。
-   Profile Info: 
    -   可以查看订阅的使用情况和过期时间 (精确到秒)。
    -   可以查看 meta-rules-dat 的相关文件的更新时间, 以及是否更新到最新的状态。
    -   显示一些比较有用的信息。
-   添加 `No Proxy Provider` 选项: 其实是将 `proxy-providers` 合并到一个文件。这样做的好处:
    -   可以解决 `proxy-providers` 的之间的节点名称相同的问题。同时 clashtui 可以使用 cron 更新, 所以并没有牺牲定时更新的功能。同时还解决了 cron 的 clashtui 进程和 mihomo 进程同时读写相同文件的竞争。
    -   一些 clash to singbox 的转换工具没有支持转换 `proxy-providers` 里的节点，所以这样更加方便转换。
    -   实现并不复杂, 详情请看 [ref](https://github.com/JohanChane/clashtui/discussions/30)。

代码相关:
-   `data.yaml`: 我的想法是, 存储需要保存到磁盘的数据。之前记录当前配置到用户的配置文件, 我觉得不是很好。之后的 clashtui 的一些需要保存的磁盘的选项也可以存储到该文件。
-   实现 profile 功能的代码有很多, 将一些与功能直接相关的接口 (比如 `update_profile`, `select_profile`, `crt_yaml_with_template`) 放置到单独的 `impl ClashTuiUtil` 以后会方便查找。
-   `download_profile` 添加 `with_proxy` 参数。当下载多个文件时, 不用每次下载文件都检查是否需要使用代理, 检查一次就够了。
-   有一次使用时, 发现 `popup_list_msg` 打印 0 行字时, 会弹出一个弹窗。所以我将其修改为, 如果是 0 行字时，则不弹出弹窗。
-   clashtui 版本号: 应该包含版本信息和 git commit (可以确定用户使用的 commit)。当 clone git 仓库时, 如果没有 tag, 则 `git describe --tag --always` 没有版本信号, 所以现在使用 `cargo_pkg_version` 和 git commit。

你还有什么疑问, 可以在这里问我, 或者用邮箱交流 (Cargo.toml 有我邮箱)。

**Comment by Jackhr-arch** [comment]
## About extra infos
Some common infos can be placed under `InfoPopUp`, while other profile-related can use `msgpopup`

## About `data.yaml`
I rewrite config.rs, and now the `config.yaml` look like 
```yaml
basic:
  clash_cfg_dir: ''
  clash_bin_pth: ''
  clash_cfg_pth: ''
  timeout: null
service: {} // Change with platform
extra:
  edit_cmd: ''
  open_dir_cmd: ''
```
and the `clashtui.conf`(renamed from `data.yaml`) looks like
```yaml
current_profile: ''
profiles: {}
```
However, these configs are still flattened in `Config`, use them as it was.

## About `download_profile` with proxy
Now I use an env var `CLASHTUI_WITH_PROXY` to control. With `CLASHTUI_WITH_PROXY` eq true, mihomo check and connectivity check will be preformed and decide whether program should upgrade profile with proxy. 

## About `popup_list_msg` printed nothing
I prefer to find out where this happened and solve it there, rather change the behavior. If you located where it is, tell me so I can get it fixed.  

## About `timeout`
Now they can be loaded from `config.yaml`, and fall back to default(5).** NOTICE**: setting this to 0 won't disable timeout, because I believe a long time(such as 30,180 or so) is enough.

## About Version code
Fantastic idea.

**Reply by JohanChane** [reply]
> ## About extra infos
> 
> Some common infos can be placed under `InfoPopUp`, while other profile-related can use `msgpopup`
> ## About `data.yaml`
> 
> I rewrite config.rs, and now the `config.yaml` look like
> 
> ```yaml
> basic:
>   clash_cfg_dir: ''
>   clash_bin_pth: ''
>   clash_cfg_pth: ''
>   timeout: null
> service: {} // Change with platform
> extra:
>   edit_cmd: ''
>   open_dir_cmd: ''
> ```
> 
> and the `clashtui.conf`(renamed from `data.yaml`) looks like
> 
> ```yaml
> current_profile: ''
> profiles: {}
> ```
> 
> However, these configs are still flattened in `Config`, use them as it was.
> ## About `download_profile` with proxy
> 
> Now I use an env var `CLASHTUI_WITH_PROXY` to control. With `CLASHTUI_WITH_PROXY` eq true, mihomo check and connectivity check will be preformed and decide whether program should upgrade profile with proxy.
> ## About `popup_list_msg` printed nothing
> 
> I prefer to find out where this happened and solve it there, rather change the behavior. If you located where it is, tell me so I can get it fixed.
> ## About `timeout`
> 
> Now they can be loaded from `config.yaml`, and fall back to default(5).** NOTICE**: setting this to 0 won't disable timeout, because I believe a long time(such as 30,180 or so) is enough.
> ## About Version code
> 
> Fantastic idea.

已采用你的 config 做法, 不过我还是觉得将 config 和 data 的数据分开比较好, 且不修改用户的 config。See [ref](https://github.com/JohanChane/clashtui/commit/59d1d4ec2488930112156dc61f197714d39a6a53)。

**Comment by JohanChane** [comment]
为了表达更加清晰, 我继续用中文讨论。

InfoPopUp:
-   原来的设计确实不是很好, 信息分开会更加协调统一。编程会更加方便, 不会因为某个错误而无法查看信息。

config.yaml:
-   这样会更加好, 不错的设计。

data.yaml: 我之前命名为 data 确实太泛了。我只要求 clashtui 不修改用户的配置即可。还要保存一个 `No Proxy Provider` 的选项。

`popup_list_msg`:
-   不记得了, 重新安装的时候吧, 这个应该很容易重现。
-   其实不用知道之前打印 0 行之处。这个接口要解决 0 行打印的问题:
    -   之前 0 行字时, 弹出的框的高度小于文字了, 不知道显示了消息, 其实是没有消息的。
    -   觉得 0 行不打印了, 不是很好的设计。可以打印一个空白行, 或者显示"No msg"的信息。
    -   这样使用该接口之前, 不用每次都检查消息是否为空。
timeout:
-   是的, 0 表示一个默认 timeout 时会更加统一。

CLASHTUI_WITH_PROXY:
-   使用环境变量, 用户切换不方便。我觉得作为一个选项 (比如: `No Proxy Provider`) 会更加方便。

**Reply by Jackhr-arch** [reply]
Yes, I've added a dialog(`confirm_popup`) to profile tab. Side effect is that "Updating...." will be skipped.

**Reply by JohanChane** [reply]
嗯, 可以的。

**Comment by Jackhr-arch** [comment]
Current, I left extra infos unfinished, and I won't finish it. You can modify them as you wish.

Common infos are located in `clashtui/src/tui/utils/info_popup.rs`, using `Infos` to manage.
And they can be set in `clashtui/src/tui/app.rs:147`, in `Keys::AppInfo =>`.
I also pub `fetch_remote` so that we can get latest infos as we open this `InfoPopUp`. 

Profile-related infos are empty now. You may copy those from `dev` or rewrite them.

**Reply by JohanChane** [reply]
这个移动一些代码就行, 我到做就行了。

**Reply by JohanChane** [reply]
已经改为没有消息时, 也会弹窗: See [ref](https://github.com/JohanChane/clashtui/commit/042e8f138745c1b55f2e760522a2caf8df5958ec)
